_posts/2023-11-17-transactional-outbox.md

```diff
@@ -0,0 +1,295 @@
+        }
+    }
+}
+
+class OutboxPublisher(private val outbox: TransactionOutbox): EventPublisher {
+    override fun publish(event: DomainEvent) {
+        outbox.schedule(RabbitMqSender::class.java).send(event)
+    }
+}
+```
+
+The `schedule` method leads to the persisting of the task, which involves sending the event to RabbitMQ. 
+This task is stored in the outbox. 
+Then, it schedules the execution of this task for after the current transaction has been committed. 
+This ensures that the event is only sent after the transaction is successfully completed.
+
+In the context of handling potential failures during event sending after a transaction, the `TransactionalOutboxScheduler` class provides a robust fallback mechanism. 
+This class periodically executes a run method that continuously calls `transactionOutbox.flush()`.
+The `flush` method in `TransactionOutbox` is responsible for finding and retrying events that haven't been successfully executed after a certain amount of time. 
+It specifically targets events that were initially set up to be executed but haven't been completed as expected and have not exceeded a predefined number of retry attempts.
+
+```kotlin
+class TransactionalOutboxScheduler(private val transactionOutbox: TransactionOutbox) {
+    @Scheduled(fixedDelay = 2000)
+    fun run() {
+        while (transactionOutbox.flush()) {
+            // NOP
+        }
+    }
+}
+```
+
+To simulate errors in event sending, utilize the `/failing-events` endpoint. 
+This endpoint generates a `FailingEvent` and attempts to send it to RabbitMQ. 
+The `RabbitMqSender` is programmed to mimic an error by intentionally failing once for each instance of `FailingEvent`.
+
+# Conclusion
+
+As we conclude our exploration of the Transactional Outbox Pattern, 
+it's evident that this pattern is more than just a technical solution—it's 
+a strategic approach to ensure data consistency and reliability in the increasingly complex landscape of distributed systems and microservices.
+
+The key takeaway from this discussion is the importance of understanding and adapting to the nuances of distributed systems. 
+The shift from traditional SQL database architectures to distributed models, while challenging, opens up new avenues for building scalable, resilient, and flexible applications. 
+The Transactional Outbox Pattern is a testament to the innovative solutions that emerge when we rethink how data consistency and reliability should be handled in modern application development.
+
+As you implement this pattern in your projects, remember that the journey towards mastering distributed systems is ongoing. 
+The Transactional Outbox Pattern is just one of the many tools in your arsenal, equipping you to tackle the challenges and leverage the opportunities presented by distributed architectures. 
+Stay curious, continue exploring, and most importantly, keep building systems that not only meet today's demands but are also prepared for tomorrow's challenges.
```

Comment:
nice smooth ending 

I liked the article a lot, as it happens to be exactly my experience level. the previous auth articles were a little too abstract for my experience level, I think. I think it was easier for me because this article has less code and more high level explanations.